### DISCUSSION TITLE
single function which subsumes different vmapped functions

### DISCUSSION THREAD

**Discussion by SaitejaUtpala** [jax-ml/jax#12391] [Q&A] [answered]
assume we have written a function which work when a, b are single instances

```
def f(a,b):
   #some computation  
   return result
```

using **vmap** we can repeat same computation for different instances of **a** and single instance of **b** as follows

v = partial(vmap, in_axes=(0, None))
af = v(f)

now similarly,  we can repeat same computation for single instances of **a** and multiple instances of **b** as follows

v = partial(vmap, in_axes=(None, 0))
bf = v(f)

we have different functions **af**, **bf**. Is there way to combine this into single function and handle automatically ? If input is batched over **a** use  `af` and if it is batched over **b** use `bf`. 

Thanks in advance.

**Comment by mattjj** [comment, marked as answer]
(Copying my response from #12392.)

Thanks for the question!

I may not understand it though. It seems like the issue is independent of JAX and vmap, and just has to do with the caller. Shouldn't this look just like e.g.

```python
if a.ndim > 1:
  return af(a, b)
elif b.ndim > 1:
  return bf(a, b)
```

where I've just assumed you check for "if the input is batched" by checking the rank of a and b somehow.

**Comment by SaitejaUtpala** [comment]
@mattjj Thanks for response. Sure that is one way. Is there are a way to skip `if/elif` ?

Because let's say if function now $f(a, b, c, d)$ 4 arguments, I want to do similar thing and it would be bit messy.

**Reply by soraros** [reply]
I'm not sure about your particular need, but would something like the following work?

```python
def vmapped(f):
  def v(*args):
    in_axes = tuple(0 if a.ndim > 1 else None for a in args)
    return vmap(f, in_axes=in_axes)(*args)
  return v
```

Or maybe you could wrap `a`, `b`, `c` in `jnp.atleast_2d` etc. before passing to the vmapped `f`.

**Reply by SaitejaUtpala** [reply]
@soraros perfect! One small question. does it effect run time in anyway ?

**Reply by soraros** [reply]
This function is traced for every combination of `ndim` for your arguments, but I won't expect the run time to be effected as long as the overall computation is jitted.
